proposals/2023/03-15_init-step.md

```diff
@@ -0,0 +1,341 @@
+# Add `InitStep` resource
+
+<!--
+The name of this markdown file should:
+
+1. Short and contain no more then 30 characters
+
+2. Contain the date of submission in MM-DD format
+
+3. Clearly state what the proposal is being submitted for
+-->
+
+| Key           | Value              |
+| :-----------: | :----------------: |
+| **Author(s)** | Jacob Floyd        |
+| **Reviewers** |                    |
+| **Date**      | March 15th, 2023   |
+| **Status**    | Waiting for Review |
+
+<!--
+If you're already working with someone, please add them to the proper author/reviewer category.
+
+If not, please leave the reviewer category empty and someone from the Vela team will assign it to themself.
+
+Here is a brief explanation of the different proposal statuses:
+
+1. Reviewed: The proposal is currently under review or has been reviewed.
+
+2. Accepted: The proposal has been accepted and is ready for implementation.
+
+3. In Progress: An accepted proposal is being implemented by actual work.
+
+NOTE: The design is subject to change during this phase.
+
+4. Cancelled: While or before implementation the proposal was cancelled.
+
+NOTE: This can happen for a multitude of reasons.
+
+5. Complete: This feature/change is implemented.
+-->
+
+## Background
+
+<!--
+This section is intended to describe the new feature, redesign or refactor.
+-->
+
+**Please provide a summary of the new feature, redesign or refactor:**
+
+<!--
+Provide your description here.
+-->
+
+Replace the injected "init" step (a pseudo-container) with `InitStep`, a new resource.
+
+Each `Build` will have one or more `InitSteps` that logically group related parts of the log data.
```

Comment:
Can you clarify how many `InitSteps` are to be expected for each `Build`?

I ask because each individual Vela installation at Cargill and Target have ran > `1 million` builds.

Each of those builds can already have multiple `services` and `steps` leading to a large amount of data in those tables as well as the `logs` table.

If we're going to now also add multiple `InitSteps` per build, and if I'm following other information in this proposal, we'll have an `InitStep` for each step/service in the build, and each `InitStep` has a corresponding entry in the `logs` table, we're talking about an exponential growth in data.

Reply:
Yes. There would be many more rows in the `InitSteps` and `Logs` tables.

If we move forward with creating an `InitStep` for each `Step` and `Service`, then:
In the logs table, the line(s) that the runtime currently prefixes to the step logs would be in a separate row from the rest of the step output.

For kubernetes, I would like to report on more than just the command to be run. Displaying the final container definition (in yaml) would be helpful as well as reporting the asynchronous events the k8s provides (pulling image, exponential back off pulling image, image pulled, container starting, container started, container terminated, crash loop back off, failed to mount volume, ...).

Today, Vela does not give me enough information about a running build, I have to use other kubernetes native tools to watch the build. So, as I started working on improving the design around init-reporting, I realized `InitStep` could be a way to first capture and then expose this additional meta state info about steps. However, I'm a kubernetes admin, and most of my Vela users will not have access to the tooling I do. So, I need to find a way to capture and expose this to those users.

As I write this, I'm trying to think of what value this might provide to docker runtime users:
- More collapsible log sections in the UI (esp in the big block of build init)
- similar to GHA and Circle CI, the command will be shown in a separate collapsible place from the rest of the command's output. And no fake shell prompt (`$`) is required.

`InitSteps` is merely one way to provide all these benefits.

With your millions of builds, how could we better tag/structure/organize the entries in the logs table without creating the exponential increase that scares you? Maybe change as more fields to the log struct so that the command(s) can be structurally separated from the output without increasing the number of rows? Turn the log struct into a postgres JSON type?